src/hotspot/share/gc/z/zMarkStackAllocator.cpp

```diff
@@ -159,13 +161,45 @@ ZMarkStackMagazine* ZMarkStackAllocator::create_magazine_from_space(uintptr_t ad
     assert(success, "Magazine should never get full");
   }
 
+  inc_nmagazine();
+
   return magazine;
 }
 
+void ZMarkStackAllocator::inc_nmagazine() {
+  int cur = Atomic::add(&_nmagazine, 1);
+  if (!ZMark::push_local_stripe() &&
+      cur * ZMarkStackMagazineSize >= ZMarkStackSpaceLimit * ZStackModeChangeRatio) {
+    ZMark::set_push_local_stripe(true);
+  }
+}
+
+void ZMarkStackAllocator::dec_nmagazine() {
+  int nmagazine = Atomic::load(&_nmagazine);
+
+  for (;;) {
+    if (nmagazine == 0) {
```

Comment:
It's confusing here _nmagazine can be negative after dec. might add some comment on this method.